diff --git a/src/dynamic-offline-data.ts b/src/dynamic-offline-data.ts
--- a/src/dynamic-offline-data.ts
+++ b/src/dynamic-offline-data.ts
@@ -140,10 +140,10 @@
   userId: string,
   type?: string,
 ): Promise<Array<T>> {
-  const filter = type ? { type, users: userId } : { users: userId };
   const db = new OfflineDb();
-  return (await db.dynamicOfflineData
-    .where(filter)
+  const collection = db.dynamicOfflineData.where('users').equals(userId);
+
+  return (await (type ? collection.and((entry) => entry.type === type) : collection)
     .toArray()
     .catch(Dexie.errnames.DatabaseClosed, () => [])) as Array<T>;
 }
@@ -216,11 +216,10 @@
     .get({
       type,
       identifier,
-      users: userId,
     })
     .catch(Dexie.errnames.DatabaseClosed, () => undefined);
 
-  if (existingEntry) {
+  if (existingEntry?.users.includes(userId)) {
     if (existingEntry.users.length > 1) {
       await db.dynamicOfflineData
         .update(existingEntry.id!, {
diff --git a/src/sync.ts b/src/sync.ts
--- a/src/sync.ts
+++ b/src/sync.ts
@@ -175,10 +175,14 @@
   const contents: Array<unknown> = [];
   const userId = await getUserId();
 
-  await db.syncQueue.where({ type, userId }).each((item, cursor) => {
-    items.push([cursor.primaryKey, item.content, item.descriptor]);
-    contents.push(item.content);
-  });
+  await db.syncQueue
+    .where('userId')
+    .equals(userId)
+    .and((item) => item.type === type)
+    .each((item, cursor) => {
+      items.push([cursor.primaryKey, item.content, item.descriptor]);
+      contents.push(item.content);
+    });
 
   for (let i = 0; i < items.length; i++) {
     const [key, item, { id, dependencies = [] }] = items[i];
@@ -233,8 +237,9 @@
   if (targetId !== undefined) {
     // in case of replacement (i.e., used same ID) we just remove the existing item
     await db.syncQueue
-      .where({ type, userId })
-      .filter((item) => item?.descriptor.id === targetId)
+      .where('userId')
+      .equals(userId)
+      .and((item) => item.type === type && item?.descriptor.id === targetId)
       .delete()
       .catch(Dexie.errnames.DatabaseClosed);
   }
@@ -279,9 +284,9 @@
  * @param type The identifying type of the synchronization items to be returned..
  */
 export async function getFullSynchronizationItemsFor<T>(userId: string, type?: string): Promise<Array<SyncItem<T>>> {
-  const filter = type ? { type, userId } : { userId };
-  return await db.syncQueue
-    .where(filter)
+  const collection = db.syncQueue.where('userId').equals(userId);
+
+  return await (type ? collection.and((item) => item.type === type) : collection)
     .toArray()
     .catch(Dexie.errnames.DatabaseClosed, () => []);
 }
